hibernate-core/src/main/java/org/hibernate/resource/jdbc/internal/LogicalConnectionManagedImpl.java

```diff
@@ -194,16 +194,28 @@ public void manualReconnect(Connection suppliedConnection) {
 	}
 
 	private void releaseConnection() {
-		//Some managed containers might trigger this release concurrently:
-		//this is not how they should do things, still we make a local
-		//copy of the variable to prevent confusing errors due to a race conditions
-		//(to trigger a more clear error, if any).
 		final Connection localVariableConnection = this.physicalConnection;
 		if ( localVariableConnection == null ) {
 			return;
 		}
 
+		// We need to set the connection to null before we release resources,
+		// in order to prevent recursion into this method.
+		// Recursion can happen when we release resources and when batch statements are in progress:
+		// when releasing resources, we'll abort the batch statement,
+		// which will trigger "logicalConnection.afterStatement()",
+		// which in some configurations will release the connection.
+
+		//Some managed containers might trigger this release concurrently:
+		//this is not how they should do things, still we try to detect it to trigger a more clear error.
+		boolean concurrentUsageDetected = ( this.physicalConnection == null );
+		this.physicalConnection = null;
+		if ( concurrentUsageDetected ) {
+			throw new HibernateException( "Detected concurrent management of connection resources." +
```

Comment:
I guess it's now impossible to ever get into this state?

I'd be OK with just removing the exception; in the original version before the "attempt to detect" we actually had an NPE triggered - that doesn't seem a likely outcome today so feel free to simplify this a bit.

Reply:
I suppose we could end up in this state in some cases since the read and write operations on `this.physicalConnection` are not atomic, but I'm not sure. In any case, I can imagine ending up in that state, but yes that sounds unlikely. Even concurrent execution of the code below sounds quite unlikely, though possible.

We could be absolutely sure concurrent execution never happens by using an `AtomicReference` for `this.physicalConnection`, but that feels like overkill, considering people are not supposed to release connections concurrently in the first place?

Anyway, I added a commit to remove this check.

Reply:
thanks :)

And yes consider that this field isn't even `volatile`, compiled code is most likely reading it just once - or even if it did twice, it's not having any memory barrier. So even the very little window which seems present is unlikely to be a real possibility.